play-route/src/main/scala/org/http4s/server/play/PlayRouteBuilder.scala

```diff
@@ -0,0 +1,149 @@
+      uri = Uri(path = requestHeader.uri),
+      headers = Headers.apply(requestHeader.headers.toMap.toList.flatMap {
+        case (headerName, values) =>
+          values.map { value =>
+            Header(headerName, value)
+          }
+      }),
+      body = EmptyBody
+    )
+
+  type ResponseStream = fs2.Stream[F, Byte]
+
+  def convertStream(responseStream: ResponseStream): PlayTargetStream = {
+    val entityBody: fs2.Stream[F, Byte] = responseStream
+    Source
+      .fromPublisher(entityBody.toUnicastPublisher())
+      .map(byte => ByteString(byte))
+  }
+
+  /**
+    * A Play accumulator Sinks HTTP data in, and then pumps out a future of a Result.
+    * That Result will have a Source as the response HTTP Entity.
+    *
+    * Here we create a unattached sink, map its materialized value into a publisher,
+    * convert that into an FS2 Stream, then pipe the request body into the http4s request.
+    */
+  def playRequestToPlayResponse(requestHeader: RequestHeader): PlayAccumulator = {
+    val sink: Sink[ByteString, Future[Result]] = {
+      Sink.asPublisher[ByteString](false).mapMaterializedValue { publisher =>
+        val requestBodyStream: fs2.Stream[F, Byte] =
+          publisher.toStream().flatMap(bs => fs2.Stream.fromIterator[F, Byte](bs.toIterator))
+
+        val wrappedResponse: F[Response[F]] =
+          F.map(unwrappedRun(
+            requestHeaderToRequest(requestHeader).withBodyStream(requestBodyStream)).value)(_.get)
+        val wrappedResult: F[Result] = F.map(wrappedResponse) { response =>
+          Result(
+            header = convertResponseToHeader(response),
+            body = Streamed(
+              data = convertStream(response.body),
+              contentLength = response.contentLength,
+              contentType = response.contentType.map(_.value)
+            )
+          )
+        }
+
+        val promise = Promise[Result]
+
+        F.runAsync(wrappedResult) {
+            case Left(bad) =>
+              promise.failure(bad)
+              IO.unit
+            case Right(good) =>
+              promise.success(good)
+              IO.unit
+          }
+          .unsafeRunSync()
+
+        promise.future
+      }
+    }
+    Accumulator.apply(sink)
+  }
+
+  def convertResponseToHeader(response: Response[F]): ResponseHeader =
+    ResponseHeader(
+      status = response.status.code,
+      headers = response.headers.collect {
+        case header
+            if !PlayRouteBuilder.AkkaHttpSetsSeparately.contains(header.parsed.name.value) =>
+          header.parsed.name.value -> header.parsed.value
+      }.toMap
+    )
+
+  /** Big big hack :-( Teach me! **/
+  def routeMatches(requestHeader: RequestHeader): Boolean = {
+    val optionalResponse: OptionT[F, Response[F]] =
+      unwrappedRun.apply(requestHeaderToRequest(requestHeader))
+    val efff: F[Option[Response[F]]] = optionalResponse.value
+    val completion = Promise[Boolean]()
+    F.runAsync(efff) {
```

Comment:
This really feels like could could just do `F.runsAsync.unsafetoFuture` and just return that,.

Reply:
`  def runAsync[A](fa: F[A])(cb: Either[Throwable, A] => IO[Unit]): IO[Unit]` so not sure I can.